### PR TITLE
DEV: Redo `DiscourseLogstashLogger` to not rely on `logstash-logger`

### PR DESCRIPTION
**Pull request by tgxworld**
This commit rewrites `DiscourseLogstashLogger` to not be an instance
of `LogstashLogger`. The reason we don't want it to be an instance of
`LogstashLogger` is because we want the new logger to be chained to
Logster's logger which can then pass down useful information like the
request's env and error backtraces which Logster has already gathered.

Note that this commit does not bother to maintain backwards
compatibility and drops the `LOGSTASH_URI` and `UNICORN_LOGSTASH_URI`
ENV variables which were previously used to configure the destination in
which `logstash-logger` would send the logs to. Instead, we introduce
the `ENABLE_LOGSTASH_LOGGER` ENV variable to replace both ENV and remove
the need for the log paths to be specified. Note that the previous
feature was considered experimental as stated in d888d3c54c4744d52b9d21d3728f5d6dbc132cde
and the new feature should be considered experimental as well. The code
may be moved into a plugin in the future.

### Reviewer notes

Unit tests have been written for `DiscourseLogstashLogger` but it is hard to test the actual content of logs being written. As such, the following manual testing instruction has been provided:

1. `cd` into your Discourse git directory
1. Run `tail -f log/development.log`
1. Run `tail -f log/unicorn.stderr.log`
1. Run `RAILS_DISABLE_ACTIVERECORD_LOGS=1 RAILS_LOGS_STDOUT=0 ENABLE_LOGSTASH_LOGGER=1 ENABLE_LOGRAGE=1 RAILS_VERBOSE_QUERY_LOGS=0  RAILS_DISABLE_ACTIVERECORD_LOGS=1 bin/rails s`
1. Look at the tail of logs to ensure they are in the JSON formatted and fields look like what they should be. 

#### Sample Rails application INFO logs

```
{"message":"[200] GET /service-worker.js (StaticController#service_worker_asset)","severity":1,"severity_name":"INFO","pid":82024,"type":"rails","host":"Guos-MBP","git_version":"7eb190425772542500a17618374570e3e0cc8837","method":"GET","path":"/service-worker.js","format":"*/*","controller":"StaticController","action":"service_worker_asset","status":200,"allocations":5569,"duration":9.5,"view":1.03,"db":0.0,"params":"","database":"default","ip":"127.0.0.1","username":null,"@timestamp":"2024-07-01T10:06:54.957Z","@version":"1"}
```

#### Sample Rails application ERROR logs

```
{"message":"NameError (Could not render layout: uninitialized constant ApplicationController::SiteSettin)\napp/controllers/application_controller.rb:65:in `has_escaped_fragment?'\napp/controllers/application_controller.rb:78:in `use_crawler_layout?'\napp/controllers/application_controller.rb:122:in `set_layout'\nlib/topic_list_responder.rb:15:in `block (2 levels) in respond_with_list'\nlib/topic_list_responder.rb:8:in `respond_with_list'\napp/controllers/list_controller.rb:111:in `block (2 levels) in \u003cclass:ListController\u003e'\napp/controllers/application_controller.rb:424:in `block in with_resolved_locale'\napp/controllers/application_controller.rb:424:in `with_resolved_locale'\nlib/middleware/omniauth_bypass_middleware.rb:64:in `call'\nlib/content_security_policy/middleware.rb:12:in `call'\nlib/middleware/csp_script_nonce_injector.rb:12:in `call'\nconfig/initializers/008-rack-cors.rb:14:in `call'\nconfig/initializers/100-quiet_logger.rb:20:in `call'\nconfig/initializers/100-silence_logger.rb:29:in `call'\nlib/middleware/missing_avatars.rb:22:in `call'\nlib/middleware/turbo_dev.rb:31:in `call'","severity":4,"severity_name":"FATAL","pid":91921,"type":"rails","host":"Guos-MBP","git_version":"8cf786fc706240860818cd755323daac46f9b0c5","backtrace":"app/controllers/application_controller.rb:65:in `has_escaped_fragment?'\napp/controllers/application_controller.rb:78:in `use_crawler_layout?'\napp/controllers/application_controller.rb:122:in `set_layout'\nactionview (7.0.8.4) lib/action_view/layouts.rb:330:in `_layout'\nactionview (7.0.8.4) lib/action_view/layouts.rb:417:in `_default_layout'\nactionview (7.0.8.4) lib/action_view/layouts.rb:393:in `block in _layout_for_option'\nactionview (7.0.8.4) lib/action_view/renderer/template_renderer.rb:108:in `resolve_layout'\nactionview (7.0.8.4) lib/action_view/renderer/template_renderer.rb:88:in `find_layout'\nactionview (7.0.8.4) lib/action_view/renderer/template_renderer.rb:71:in `render_with_layout'\nactionview (7.0.8.4) lib/action_view/renderer/template_renderer.rb:59:in `render_template'\nactionview (7.0.8.4) lib/action_view/renderer/template_renderer.rb:11:in `render'\nactionview (7.0.8.4) lib/action_view/renderer/renderer.rb:61:in `render_template_to_object'\nactionview (7.0.8.4) lib/action_view/renderer/renderer.rb:29:in `render_to_object'\nactionview (7.0.8.4) lib/action_view/rendering.rb:117:in `block in _render_template'\nactionview (7.0.8.4) lib/action_view/base.rb:270:in `in_rendering_context'\nactionview (7.0.8.4) lib/action_view/rendering.rb:116:in `_render_template'\nactionpack (7.0.8.4) lib/action_controller/metal/streaming.rb:216:in `_render_template'\nactionview (7.0.8.4) lib/action_view/rendering.rb:103:in `render_to_body'\nactionpack (7.0.8.4) lib/action_controller/metal/rendering.rb:158:in `render_to_body'\nactionpack (7.0.8.4) lib/action_controller/metal/renderers.rb:141:in `render_to_body'\nactionpack (7.0.8.4) lib/abstract_controller/rendering.rb:27:in `render'\nactionpack (7.0.8.4) lib/action_controller/metal/rendering.rb:139:in `render'\nactionpack (7.0.8.4) lib/action_controller/metal/instrumentation.rb:22:in `block (2 levels) in render'\n/Users/tgxworld/.asdf/installs/ruby/3.3.2/lib/ruby/3.3.0/benchmark.rb:313:in `realtime'\nactivesupport (7.0.8.4) lib/active_support/core_ext/benchmark.rb:14:in `ms'\nactionpack (7.0.8.4) lib/action_controller/metal/instrumentation.rb:22:in `block in render'\nactionpack (7.0.8.4) lib/action_controller/metal/instrumentation.rb:91:in `cleanup_view_runtime'\nactiverecord (7.0.8.4) lib/active_record/railties/controller_runtime.rb:34:in `cleanup_view_runtime'\nactionpack (7.0.8.4) lib/action_controller/metal/instrumentation.rb:21:in `render'\nlib/topic_list_responder.rb:15:in `block (2 levels) in respond_with_list'\nactionpack (7.0.8.4) lib/action_controller/metal/mime_responds.rb:214:in `respond_to'\nlib/topic_list_responder.rb:8:in `respond_with_list'\napp/controllers/list_controller.rb:111:in `block (2 levels) in \u003cclass:ListController\u003e'\nactionpack (7.0.8.4) lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'\nactionpack (7.0.8.4) lib/abstract_controller/base.rb:215:in `process_action'\nactionpack (7.0.8.4) lib/action_controller/metal/rendering.rb:165:in `process_action'\nactionpack (7.0.8.4) lib/abstract_controller/callbacks.rb:234:in `block in process_action'\nactivesupport (7.0.8.4) lib/active_support/callbacks.rb:118:in `block in run_callbacks'\napp/controllers/application_controller.rb:424:in `block in with_resolved_locale'\ni18n (1.14.5) lib/i18n.rb:351:in `with_locale'\napp/controllers/application_controller.rb:424:in `with_resolved_locale'\nactivesupport (7.0.8.4) lib/active_support/callbacks.rb:127:in `block in run_callbacks'\nactivesupport (7.0.8.4) lib/active_support/callbacks.rb:138:in `run_callbacks'\nactionpack (7.0.8.4) lib/abstract_controller/callbacks.rb:233:in `process_action'\nactionpack (7.0.8.4) lib/action_controller/metal/rescue.rb:23:in `process_action'\nactionpack (7.0.8.4) lib/action_controller/metal/instrumentation.rb:67:in `block in process_action'\nactivesupport (7.0.8.4) lib/active_support/notifications.rb:206:in `block in instrument'\nactivesupport (7.0.8.4) lib/active_support/notifications/instrumenter.rb:24:in `instrument'\nactivesupport (7.0.8.4) lib/active_support/notifications.rb:206:in `instrument'\nactionpack (7.0.8.4) lib/action_controller/metal/instrumentation.rb:66:in `process_action'\nactionpack (7.0.8.4) lib/action_controller/metal/params_wrapper.rb:259:in `process_action'\nactiverecord (7.0.8.4) lib/active_record/railties/controller_runtime.rb:27:in `process_action'\nactionpack (7.0.8.4) lib/abstract_controller/base.rb:151:in `process'\nactionview (7.0.8.4) lib/action_view/rendering.rb:39:in `process'\nrack-mini-profiler (3.3.1) lib/mini_profiler/profiling_methods.rb:115:in `block in profile_method'\nactionpack (7.0.8.4) lib/action_controller/metal.rb:188:in `dispatch'\nactionpack (7.0.8.4) lib/action_controller/metal.rb:251:in `dispatch'\nactionpack (7.0.8.4) lib/action_dispatch/routing/route_set.rb:49:in `dispatch'\nactionpack (7.0.8.4) lib/action_dispatch/routing/route_set.rb:32:in `serve'\nactionpack (7.0.8.4) lib/action_dispatch/routing/mapper.rb:18:in `block in \u003cclass:Constraints\u003e'\nactionpack (7.0.8.4) lib/action_dispatch/routing/mapper.rb:48:in `serve'\nactionpack (7.0.8.4) lib/action_dispatch/journey/router.rb:50:in `block in serve'\nactionpack (7.0.8.4) lib/action_dispatch/journey/router.rb:32:in `each'\nactionpack (7.0.8.4) lib/action_dispatch/journey/router.rb:32:in `serve'\nactionpack (7.0.8.4) lib/action_dispatch/routing/route_set.rb:852:in `call'\nlib/middleware/omniauth_bypass_middleware.rb:64:in `call'\nrack (2.2.9) lib/rack/tempfile_reaper.rb:15:in `call'\nrack (2.2.9) lib/rack/conditional_get.rb:27:in `call'\nrack (2.2.9) lib/rack/head.rb:12:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/http/permissions_policy.rb:38:in `call'\nlib/content_security_policy/middleware.rb:12:in `call'\nlib/middleware/csp_script_nonce_injector.rb:12:in `call'\nconfig/initializers/008-rack-cors.rb:14:in `call'\nrack (2.2.9) lib/rack/session/abstract/id.rb:266:in `context'\nrack (2.2.9) lib/rack/session/abstract/id.rb:260:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/cookies.rb:704:in `call'\nactiverecord (7.0.8.4) lib/active_record/migration.rb:638:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'\nactivesupport (7.0.8.4) lib/active_support/callbacks.rb:99:in `run_callbacks'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/callbacks.rb:26:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/executor.rb:14:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/actionable_exceptions.rb:17:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/debug_exceptions.rb:28:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/show_exceptions.rb:29:in `call'\n/Users/tgxworld/work/logster/lib/logster/middleware/reporter.rb:40:in `call'\nlograge (0.14.0) lib/lograge/rails_ext/rack/logger.rb:18:in `call_app'\nrailties (7.0.8.4) lib/rails/rack/logger.rb:27:in `call'\nconfig/initializers/100-quiet_logger.rb:20:in `call'\nconfig/initializers/100-silence_logger.rb:29:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/remote_ip.rb:93:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/request_id.rb:26:in `call'\nrack (2.2.9) lib/rack/method_override.rb:24:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/executor.rb:14:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/static.rb:23:in `call'\nrack (2.2.9) lib/rack/sendfile.rb:110:in `call'\nactionpack (7.0.8.4) lib/action_dispatch/middleware/host_authorization.rb:138:in `call'\nlib/middleware/missing_avatars.rb:22:in `call'\nlib/middleware/turbo_dev.rb:31:in `call'\nrack-mini-profiler (3.3.1) lib/mini_profiler.rb:334:in `call'\nmessage_bus (4.3.8) lib/message_bus/rack/middleware.rb:60:in `call'\nrailties (7.0.8.4) lib/rails/engine.rb:530:in `call'\nrailties (7.0.8.4) lib/rails/railtie.rb:226:in `public_send'\nrailties (7.0.8.4) lib/rails/railtie.rb:226:in `method_missing'\nrack (2.2.9) lib/rack/urlmap.rb:74:in `block in call'\nrack (2.2.9) lib/rack/urlmap.rb:58:in `each'\nrack (2.2.9) lib/rack/urlmap.rb:58:in `call'\nunicorn (6.1.0) lib/unicorn/http_server.rb:634:in `process_client'\nunicorn (6.1.0) lib/unicorn/http_server.rb:739:in `worker_loop'\nunicorn (6.1.0) lib/unicorn/http_server.rb:547:in `spawn_missing_workers'\nunicorn (6.1.0) lib/unicorn/http_server.rb:143:in `start'\nunicorn (6.1.0) bin/unicorn:128:in `\u003ctop (required)\u003e'\nbin/unicorn:93:in `load'\nbin/unicorn:93:in `block in \u003cmain\u003e'\nbin/unicorn:92:in `fork'\nbin/unicorn:92:in `\u003cmain\u003e'","request.headers.request_uri":"/","request.headers.request_method":"GET","request.headers.http_host":"127.0.0.1:3000","request.headers.http_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36","request.headers.http_accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7","request.headers.http_referer":null,"request.headers.http_x_forwarded_for":null,"request.headers.http_x_real_ip":null,"database":"default"}
``` 

#### Sample Unicorn logs output

```
{"message":"master complete","severity":1,"severity_name":"INFO","pid":82024,"type":"unicorn","host":"Guos-MBP","git_version":"7eb190425772542500a17618374570e3e0cc8837"}
```

### CODE REVIEW

File: config/environments/development.rb [hunk 1]
```diff
@@ -90,8 +91,6 @@
       end
     end
 
-    ActiveRecord::Base.logger = nil if ENV["RAILS_DISABLE_ACTIVERECORD_LOGS"] == "1"
```
Comment (tgxworld) [inline config/environments/development.rb:93]: Not sure why this was setting `ActiveRecord::Base` directly in the `after_initialize` block when there is `config.active_record.logger`

File: config/initializers/101-lograge.rb [hunk 2]
```diff
@@ -101,18 +101,25 @@
...
+                event["database"] = RailsMultisite::ConnectionManagement.current_db
+              end,
+          )
 
         # Remove ActiveSupport::Logger from the chain and replace with Lograge's
-        # logger
-        Rails.logger.chained.pop
-        Rails.logger.chain(config.lograge.logger)
+        # logger.
+        Logster.logger.instance_variable_get(:@chained).delete(Logster.logger.chained.first) # Reach into instance variable but this will be irrelevant in Rails 7.1
```
Comment (tgxworld) [inline config/initializers/101-lograge.rb:121]: In Rails 7.1, the `ActiveSupport::Logger` will no longer be chained to Logster.

File: lib/discourse.rb [hunk 3]
```diff
@@ -828,45 +831,6 @@ def self.request_refresh!(user_ids: nil)
...
-  end
-
-  def self.try_git(git_cmd, default_value)
-    begin
-      `#{git_cmd}`.strip
-    rescue StandardError
-      default_value
-    end.presence || default_value
-  end
-
```
Comment (tgxworld) [inline lib/discourse.rb:869]: Extracted this to a module that can be used by `DiscourseLogstashLogger`

File: lib/discourse_logstash_logger.rb [hunk 4]
```diff
@@ -1,18 +1,103 @@
...
+  ALLOWED_HEADERS_FROM_ENV = %w[
+    REQUEST_URI
+    REQUEST_METHOD
+    HTTP_HOST
+    HTTP_USER_AGENT
+    HTTP_ACCEPT
+    HTTP_REFERER
+    HTTP_X_FORWARDED_FOR
+    HTTP_X_REAL_IP
+  ]
```
Comment (tgxworld) [inline lib/discourse_logstash_logger.rb:46]: Copied this [from Logster](https://github.com/discourse/logster/blob/85c05539e274e0ba2e30411409ee181f462d5f43/lib/logster/message.rb#L9-L22) but we don't have to keep this in sync. Our app might want to log more stuff.

File: lib/discourse_logstash_logger.rb [hunk 5]
```diff
@@ -1,18 +1,103 @@
...
+  def self.logger(logdev:, type:, customize_event: nil)
+    logger = self.new(logdev)
+    logger.type = type
+    logger.customize_event = customize_event if customize_event
+    logger
+  end
+
+  def add(*args, &block)
+    add_with_opts(*args, &block)
+  end
```
Comment (tgxworld) [inline lib/discourse_logstash_logger.rb:35]: This is overriding the `Logger#add` method and is a pattern that `Logster` uses. Not the most ideal but I don't have a better idea.

File: lib/discourse_logstash_logger.rb [hunk 6]
```diff
@@ -1,18 +1,103 @@
...
+    progname = @progname if progname.nil?
+
+    if message.nil?
+      if block_given?
+        message = yield
+      else
+        message = progname
+        progname = @progname
+      end
+    end
```
Comment (tgxworld) [inline lib/discourse_logstash_logger.rb:60]: This is what the `Logger#add` method does.

File: lib/discourse_logstash_logger.rb [hunk 7]
```diff
@@ -1,18 +1,103 @@
...
+      end
+
+      if (env = opts&.dig(:env)).present?
+        ALLOWED_HEADERS_FROM_ENV.each do |header|
+          event["request.headers.#{header.downcase}"] = opts[:env][header]
+        end
+      end
+    end
+
+    if message.start_with?("{") && message.end_with?("}")
```
Comment (tgxworld) [inline lib/discourse_logstash_logger.rb:86]: This looks abit naive but is actually what `logstash-logger` was doing previously. Not going to change what works.

File: lib/version.rb [hunk 8]
```diff
@@ -16,8 +15,8 @@ module VERSION #:nodoc:
       MAJOR = PARTS[0].to_i
       MINOR = PARTS[1].to_i
       TINY = PARTS[2].to_i
-      PRE = PARTS[3]&.split("-", 2)&.first
-      DEV = PARTS[3]&.split("-", 2)&.second
```
Comment (tgxworld) [inline lib/version.rb:20]: `Array#first` and `Array#second` are Rails methods but this module may be used before Rails is loaded.

File: lib/discourse_logstash_logger.rb [hunk 9]
```diff
@@ -1,18 +1,103 @@
...
+    if severity >= Logger::WARN
+      if (backtrace = opts&.dig(:backtrace)).present?
+        event["backtrace"] = backtrace
+      end
+
+      if (env = opts&.dig(:env)).present?
+        ALLOWED_HEADERS_FROM_ENV.each do |header|
+          event["request.headers.#{header.downcase}"] = opts[:env][header]
+        end
+      end
```
Comment (tgxworld) [inline lib/discourse_logstash_logger.rb:83]: Note that the `env` and `backtrace` keys in opts are passed in by Logster to loggers that has been "chained" to it via the `Logster::Logger#chain` API. Basically, the instance of `Logster::Logger` will forward the logs to the loggers that have been chained. 

The change in https://github.com/discourse/logster/pull/238 updates `Logster::Logger` such that `Logster::Logger` will pass its `opts` to the chained logger if `with_opts: true` was set when calling `Logster::Logger#chain`

File: config/initializers/101-lograge.rb (hunk 2 above)
Comment (nattsw) [reply config/initializers/101-lograge.rb:121]: https://github.com/discourse/logster/pull/238/files#r1660938639

**Review by nbianca — APPROVED** [review]

**Pull request merged**